var/spack/repos/builtin/packages/py-pykokkos-base/package.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+#
+# ----------------------------------------------------------------------------
+
+from spack import *
+
+
+class PyPykokkosBase(CMakePackage, PythonPackage):
+    """Minimal set of bindings for Kokkos interoperability with Python
+    (initialize, finalize, View, DynRankView, Kokkos-tools)"""
+
+    homepage = "https://github.com/kokkos/pykokkos-base.git"
+    git = "https://github.com/kokkos/pykokkos-base.git"
+    maintainers = ["jrmadsen"]
+
+    version("main", branch="main", submodules=False)
+    version(
+        "0.0.5", commit="45f6e892c007ab124fabb3a545f4744537eafb55", submodules=False
+    )
+    version(
+        "0.0.4", commit="2efe1220d0128d3f2d371c9ed5234c4978d73a77", submodules=False
+    )
+    version(
+        "0.0.3", commit="4fe4421ac624ba2efe1eee265153e690622a18a5", submodules=False
+    )
+
+    variant(
+        "layouts",
+        default=True,
+        description="Build Kokkos View/DynRankView with layout variants",
+    )
+    variant(
+        "memory_traits",
+        default=True,
+        description="Build Kokkos View/DynRankView with memory trait variants",
+    )
+    variant(
+        "view_ranks",
+        default="4",
+        description="Max Kokkos View dimensions",
+        values=("1", "2", "3", "4", "5", "6", "7"),
+        multi=False,
+    )
+
+    extends("python")
+    depends_on("cmake@3.16:", type="build")
+    depends_on("py-pybind11", type="build")
+    depends_on("kokkos@3.4.00:", type=("build", "run"))
+    depends_on("python@3:", type=("build", "run"))
+
+    def cmake_args(self):
+        spec = self.spec
+
+        args = [
+            "-DENABLE_INTERNAL_KOKKOS=OFF",
+            "-DENABLE_INTERNAL_PYBIND11=OFF",
+            "-DPYTHON_EXECUTABLE={0}".format(spec["python"].command.path),
+            "-DPython3_EXECUTABLE={0}".format(spec["python"].command.path),
+            "-DENABLE_VIEW_RANKS={0}".format(spec.variants["view_ranks"].value),
```

Comment:
There are CMakePackage helper functions to simplify construction arguments, and stylistically Spack packages tend to use single quotes instead of double:

```suggestion
            self.define('ENABLE_INTERNAL_KOKKOS', False),
            self.define('ENABLE_INTERNAL_PYBIND11', False),
            self.define('PYTHON_EXECUTABLE', spec['python'].command.path),
            self.define('Python3_EXECUTABLE', spec['python'].command.path),
            self.define_from_variant('ENABLE_VIEW_RANKS', 'view_ranks'),
```

Reply:
So you are saying I have to go in and change every double quotes to single quotes? Is there an auto-formatter? Black doesn't have an option for single quotes. Enforcing stylistic standards without an auto-formatter is a super annoying practice.

Reply:
I'm not saying that, no—it's  just a heads-up for future new packages. (Not that `s/"/'/g` is too much trouble 😉 ) The main change here is using the helper functions for `define` and `define_from_variant`, which come with the extra benefit of automatically validating the existence of variant names (for `_from_variant`) and standardizing the usage across packages.

Reply:
running sed once isn't too much trouble. It's having to run it after every time I run black that is an issue.

Reply:
Plus single-quote strings offend my C/C++ sensibilities

Reply:
What, black *mandates* double quotes??

Reply:
Yep, there's an option to not change strings but there is no "use single quotes instead of double quotes".

Reply:
Interesting. I always used double-quotes in Python as well (having also come from a C++ background) until I noticed that Spack and other major packages (Numpy, Matplotlib) tended to use single quotes. It's very interesting that Black tries to force double-quotes (indeed there looks to have been [at least one heated discussion about it](https://github.com/psf/black/issues/118))... I think the decision to allow both single and double without preferring one was a mistake for the python standards.

Reply:
I'm not sure about how much bindings matplotlib has but numpy using single is strange given how much C code there is.